p2p/host/basic/basic_host.go

```diff
@@ -377,9 +390,43 @@ func (h *BasicHost) Connect(ctx context.Context, pi pstore.PeerInfo) error {
 		return nil
 	}
 
+	resolved, err := h.resolveAddrs(ctx, h.Peerstore().PeerInfo(pi.ID))
+	if err != nil {
+		return err
+	}
+	h.Peerstore().AddAddrs(pi.ID, resolved, pstore.TempAddrTTL)
+
 	return h.dialPeer(ctx, pi.ID)
 }
 
+func (h *BasicHost) resolveAddrs(ctx context.Context, pi pstore.PeerInfo) ([]ma.Multiaddr, error) {
+	proto := ma.ProtocolWithCode(ma.P_IPFS).Name
+	p2paddr, err := ma.NewMultiaddr("/" + proto + "/" + pi.ID.Pretty())
+	if err != nil {
+		return nil, err
+	}
+
+	var addrs []ma.Multiaddr
+	for _, addr := range pi.Addrs {
+		addrs = append(addrs, addr)
+
+		reqaddr := addr.Encapsulate(p2paddr)
+		resaddrs, err := h.maResolver.Resolve(ctx, reqaddr)
```

Comment:
The issue is that if it just returns its input, then the multiaddr will end up being appended twice, and the second time with needless processing. 

So it's not a noop; it both duplicates the input (the address has already been appended) and has overhead on top, as it needs to perform a few operations that eat cpu cycles.